mesh/submesh/submesh_utils.cpp

```diff
@@ -227,5 +253,607 @@ Array<int> BuildFaceMap(const Mesh& pm, const Mesh& sm,
+   std::set<int> new_nodes;
+   parent_to_submesh_element_ids.reserve(parent.GetNumFaces());
+   parent_element_ids.Reserve(parent.GetNumFaces());
+   // Base class cast then const cast because GetFaceList uses just in time
+   // construction.
+   const auto &face_list = const_cast<NCMesh&>(static_cast<const NCMesh&>
+                                               (parent)).GetFaceList();
+   // Double indexing loop because begin() and end() do not align with index 0
+   // and size-1.
+   for (int i = 0, ipe = 0; ipe < parent.GetNumFaces(); i++)
+   {
+      const auto &face = parent.GetFace(i);
+      if (face.Unused()) { continue; }
+      ipe++; // actual possible parent element.
+      if (!HasAttribute(face, attributes)
+          || face_list.GetMeshIdType(face.index) == NCMesh::NCList::MeshIdType::MASTER
+         ) { continue; }
+
+      auto fn = FaceNodes{submesh.parent_->FindFaceNodes(face)};
+      if (pnodes_new_elem.find(fn) != pnodes_new_elem.end()) { continue; }
+
+      // TODO: Internal nc submesh can be constructed and solved on, but the
+      // transfer to the parent mesh can be erroneous, this is likely due to not
+      // treating the changing orientation of internal faces for ncmesh within
+      // the ptransfermap.
+      MFEM_ASSERT(face.elem[0] < 0 || face.elem[1] < 0,
+                  "Internal nonconforming boundaries are not reliably supported yet.");
+      auto face_geom = FaceGeomFromNodes(fn.nodes);
+      int new_elem_id = submesh.AddElement(face_geom, face.attribute);
+
+      // Rank needs to be established by presence (or lack of) in the submesh.
+      submesh.elements[new_elem_id].rank = [&parent, &face]()
+      {
+         auto rank0 = face.elem[0] >= 0 ? parent.GetElement(face.elem[0]).rank : -1;
+         auto rank1 = face.elem[1] >= 0 ? parent.GetElement(face.elem[1]).rank : -1;
+         if (rank0 < 0) { return rank1; }
+         if (rank1 < 0) { return rank0; }
+         return rank0 < rank1 ? rank0 : rank1;
+      }();
+      pnodes_new_elem[fn] = new_elem_id;
+      parent_element_ids.Append(i);
+      parent_to_submesh_element_ids[i] = new_elem_id;
+
+      // Copy in the parent nodes. These will be relabeled once the tree is
+      // built.
+      std::copy(fn.nodes.begin(), fn.nodes.end(), submesh.elements[new_elem_id].node);
+      for (auto x : fn.nodes)
+         if (x != -1)
+         {
+            new_nodes.insert(x);
+         }
+      auto &gi = submesh.GI[face_geom];
+      gi.InitGeom(face_geom);
+      for (int e = 0; e < gi.ne; e++)
+      {
+         new_nodes.insert(submesh.ParentNodes().FindId(fn.nodes[gi.edges[e][0]],
+                                                       fn.nodes[gi.edges[e][1]]));
+      }
+
+      /*
+         - Check not top level face
+         - Check for parent of the newly entered element
+            - if not present, add in
+            - if present but different order and this path is non-ambiguous, reorder so consistent with child
```

Comment:
Long line.

Reply:
Fixed in https://github.com/mfem/mfem/pull/4476/commits/14e511403985758e7c9c8c131907bcad1a3a995b